### PR TITLE
fix mutation transforms

### LINKED ISSUE #71918: Transformed mutations should block gaining the transform target mutation during mutation gain process
### Describe the bug

If I have a mutation "Body Armor" that all it does it give an activation button that allows me to transform the mutation into an active mutation "Body Armor (Active)" then proceed to down mutagen while I have the active version turned on I can then mutate into "Body Armor", so now I have both traits.  This shouldn't be possible.  It's currently worked around in any mutations that give enchantment driven effects via conditionals but that's not an ideal solution until all effects are enchantment driven which is a gigantic project that has already claimed multiple attempts.

### Attach save file

N/A

### Steps to reproduce

Find a mutation that transforms, transform, prepare to mutate, rng into the inactive version, weep.

### Expected behavior

I expect that mutate into would check for and account for active versions being the passive version.

### Screenshots

_No response_

### Versions and configuration

N/A

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by worm-girl**
#### Summary
Bugfixes "Fix transformable mutations breaking prereqs and type restrictions"

#### Purpose of change

Fixes #71918 

Mutation transformation via mutation jsons was implemented a long time ago, and was causing some errors if a person drank mutagen or purifier while using some of them. It was difficult to keep on top of these with the type/category system, and a lot of the active mutation code is just very old. EOCs and triggers are much more powerful, so let's use those instead.

#### Describe the solution

Retractable Claws - Serves as a base mutation which holds all the prereqs, category, type data, etc. Activating it adds a second mutation, Retractable Claws (extended). Activating it again removes this mutation.
Normal Human - Created Alter Ego, which serves as a base mutation which holds all the prereqs, category, type data, etc. EOCs will variably apply Normal Human or Violent Monster and remove the other as appropriate. Note that Alter Ego is invisible, the player only sees Normal Human and Violent Monster, as usual.
Eagle-Eyed - You get a base mutation that never changes. You also get one of two invisible mutations that alternate based on whether it's night or day. This is a little more complex than it ought to be, but eagle-eyed night and day modes already existed and I don't want to break compatibility.
Photophore 1 and 2 - These are each a single mutation which runs an EOC that adds or removes a secondary one when activated.
Hyper-Metabolism - This is a base mutation that adds a second one, hyper-metabolism (burning) when activated. This second mutation starts active and turns itself off when it runs out of kcal, but can also be manually deactivated.
Insect Wings - This is a base mutation that adds a second one, Insect Wings (buzzing) when activated. This second mutation removes itself when you run out of stamina, but can also be manually deactivated, which removes it.
Shifting Bark - This is a single base mutation that adds 1 of 3 invisible add-ons, which each grant a different integrated armor. You can cycle through them by activating the base mutation.

The base mutations are the only ones with prereqs, types, categories, etc. The add-ons are all invisible (except for hyper-metabolism: burning, which contains important information, and normal human/violent monster, which are visible while the base is not), not purifiable, and not valid.

Purifying a base mutation gets rid of its add-ons.

There was also a minor bug with sneezing where it could happen in your sleep. While I was editing mutation EOCs, This was unintentional, I fixed it.

#### Describe alternatives you've considered

Updating the mutation.cpp stuff to allow active mutations to do a lot more, but I think EOCs are probably the desired solution in any case where they work.

#### Testing

Hyper-metabolism: Gained the mutation, activated it, deactivated it. Activated it, waited until I began starving, it deactivated itself. Activated the mutation and took purifier, the base mutation was removed by purifier, automatically removing the active version.
Alter Ego: Gained the mutation, waited until midnight. Turned into a Violent Monster, returned to normal at 6 AM. Ate human flesh, plumetting my mood down to -60, transformed immediately. Stayed transformed for a couple of hours until my mood was better. Turned into a violent monster again, took purifier, lost the base trait which also made violent monster go away.
Retractable Claws: Extended and retracted claws. Took purifier while they were extended, saw they successfully reverted to normal claws with no weirdness.
Weak Photophore and Photophore: Gained the mutations. Activated and deactivated them. Took purifier while they were on, saw that the base mutations were removed, also removing the active add-on.
Insect wings: Gained the mutation. Turned it on and off. Turned it on, purified, lost the base mutation and the invisible active add-on.
Shifting Bark: Gained the mutation, swapped between all the bark types. Purified and lost the base mutation and the invisible add-on, plus the integrated armor it granted.

### CODE REVIEW

File: tests/materials_test.cpp [hunk 1]
```diff
@@ -45,9 +45,9 @@ TEST_CASE( "Resistance_vs_material_portions", "[material]" )
     REQUIRE( mostly_plastic.get_base_material().id == material_plastic );
 
     CHECK( mostly_steel.resist( damage_cut ) > mostly_plastic.resist( damage_cut ) );
-    CHECK( mostly_steel.resist( damage_acid ) == mostly_plastic.resist( damage_acid ) );
+    CHECK( mostly_steel.resist( damage_acid ) > mostly_plastic.resist( damage_acid ) );
```
Comment (RenechCDDA) [inline tests/materials_test.cpp:48]: This change (and all others to material test) appear unrelated, and should be separated.

File: data/json/mutations/mutations.json [hunk 2]
```diff
@@ -339,25 +339,26 @@
...
+    "category": [ "ELFA", "INSECT", "FISH" ],
+    "active": true,
+    "activated_eocs": [ "EOC_BIOLUM1" ]
   },
   {
     "type": "mutation",
     "id": "BIOLUM1_active",
     "name": { "str": "Weak Photophore (on)" },
     "description": "Your photophore is glowing softly.",
-    "copy-from": "BIOLUM1",
```
Comment (RenechCDDA) [inline data/json/mutations/mutations.json:358]: This effectively erases the `BIOLUM1` mutation from any existing saves which had it acitve.

File: data/json/mutations/mutations.json [hunk 3]
```diff
@@ -3854,21 +3830,19 @@
...
+    "activated_eocs": [ "EOC_CLAWS_RETRACT" ],
     "category": [ "FELINE" ],
-    "transform": { "target": "CLAWS_RETRACT_active", "msg_transform": "You extend your claws.", "active": false, "moves": 10 },
     "cost": 0
   },
   {
     "type": "mutation",
     "id": "CLAWS_RETRACT_active",
     "name": { "str": "Extended Claws" },
-    "copy-from": "CLAWS_RETRACT",
```
Comment (RenechCDDA) [inline data/json/mutations/mutations.json:3865]: Ditto on the deletion in existing saves, and I assume elsewhere (this is not a comprehensive review)

File: tests/mutation_test.cpp [hunk 4]
```diff
@@ -25,7 +25,7 @@ static const mutation_category_id mutation_category_MOUSE( "MOUSE" );
 static const mutation_category_id mutation_category_RAPTOR( "RAPTOR" );
 static const mutation_category_id mutation_category_REMOVAL_TEST( "REMOVAL_TEST" );
 
-static const trait_id trait_EAGLEEYED( "EAGLEEYED" );
+static const trait_id trait_EAGLEEYED( "EAGLEEYED_DAY" );
```
Comment (RenechCDDA) [inline tests/mutation_test.cpp:28]:
```suggestion
static const trait_id trait_EAGLEEYED_DAY( "EAGLEEYED_DAY" );
```

File: tests/materials_test.cpp (hunk 1 above)
Comment (fairyarmadillo) [reply tests/materials_test.cpp:48]: Oops, those got in from another PR. Thanks for catching that.

File: data/json/mutations/mutations.json (hunk 2 above)
Comment (fairyarmadillo) [reply data/json/mutations/mutations.json:358]: Hm. The only thing I can think of is obsoleting BIOLUM1_active, adding an EOC that gives you BIOLUM1 and gets rid of BIOLUM1_active if you have it, and then moving the active from of BIOLUM1 to a new ID. But that seems kinda hacky and I don't think it'd get merged. Any thoughts on a better method?

Comment (RenechCDDA) [reply data/json/mutations/mutations.json:358]: You should just be able to migrate the mutation `BIOLUM1_ACTIVE` to `BIOLUM1`, the conditional enchantment will re-check on load (when the enchantment cache is dirtied, which happens many times on load) if it is necessary and apply it when the enchantment's conditions are fulfilled.

Comment (fairyarmadillo) [reply data/json/mutations/mutations.json:358]: Oh right, I forgot about migrations. Thanks for the help, I'll reopen when I get that set up.

**Comment by Maleclypse** [comment]
> @RenechCDDA I've got the migrations done, but if I give myself biolum2_active, then save, quit, and reload, I don't gain biolum1. Do migrations only affect saves that were made before the migration was added?

No they should work on every new load until they are removed.  That's part of the reason that eventually a stable or two after they are added migrations are removed.

**Comment by fairyarmadillo** [comment]
Alright, I'll keep trying to figure this out then.

**Comment by RenechCDDA** [comment]
Isn't that because you are migrating `BIOLUM2_ACTIVE` to `BIOLUM2`? If you save with BIOLUM2_ACTIVE, apply the migration, and then reload, you should end up with BIOLUM2.

File: data/json/mutations/mutations.json [hunk 5]
```diff
@@ -5427,19 +5403,20 @@
...
     "id": "WINGS_INSECT_active",
     "name": { "str": "Insect Wings (Buzzing)" },
-    "description": "You're fluttering your wings as fast as you can.  This doesn't generate enough lift to move you, but it does increase your movement speed.  If you're not burdened, you can fall short distances without harm.",
-    "copy-from": "WINGS_INSECT",
+    "description": "You're fluttering your wings as fast as you can.  This shouldn't be visible, it's a bug if it is.",
     "valid": false,
+    "purifiable": false,
+    "points": 0,
+    "player_display": false,
     "movecost_modifier": 0.75,
```
Comment (Wishbringer) [inline data/json/mutations/mutations.json:5417]:
```suggestion
```
Movecost change was moved into `ench_wings_insect_active` as an enchantment.

File: data/json/mutations/mutations.json [hunk 6]
```diff
@@ -406,54 +408,33 @@
...
-              { "math": [ "time_since('midnight', 'unit': 'hours') > 6" ] },
-              { "math": [ "time_since('midnight', 'unit': 'hours') < 21" ] }
-            ]
-          },
-          "msg_on": { "text": "As the morning comes, you return back to normal.", "rating": "good" }
-        }
-      ]
-    ],
-    "transform": { "target": "JEKYLL", "msg_transform": "The anger subsides.", "active": false, "moves": 10 },
     "passive_mods": { "int_mod": -8, "str_mod": 3 },
```
Comment (Wishbringer) [inline data/json/mutations/mutations.json:438]:
```suggestion
    "enchantments": [ { "values": [ { "value": "STRENGTH", "add": 3 }, { "value": "INTELLIGENCE", "add": -8 } ] } ],
```
